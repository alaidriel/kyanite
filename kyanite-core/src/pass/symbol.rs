use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    ops::{Deref, DerefMut},
};

use crate::{
    ast::{node, Decl, Type},
    token::Token,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Binding {
    Function(node::FuncDecl),
    Constant(node::ConstantDecl),
    Variable(node::VarDecl),
}

impl Binding {
    pub fn ty(&self) -> Type {
        match self {
            Self::Function(fun) => Type::from(fun.ty.as_ref()),
            Self::Constant(c) => Type::from(&c.ty),
            Self::Variable(v) => Type::from(&v.ty),
        }
    }
}

#[derive(Debug)]
pub struct SymbolTable(HashMap<Token, Binding>);

impl SymbolTable {
    pub fn new() -> Self {
        Self(HashMap::new())
    }
}

impl Deref for SymbolTable {
    type Target = HashMap<Token, Binding>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for SymbolTable {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl Default for SymbolTable {
    fn default() -> Self {
        Self::new()
    }
}

impl From<&Vec<Decl>> for SymbolTable {
    fn from(nodes: &Vec<Decl>) -> Self {
        let mut table = Self::new();
        for node in nodes {
            self::SymbolTableVisitor::visit(node, &mut table);
        }
        table
    }
}

trait SymbolTableVisitor {
    fn visit(&self, table: &mut SymbolTable);
}

impl SymbolTableVisitor for Decl {
    fn visit(&self, table: &mut SymbolTable) {
        match self {
            Decl::Function(fun) => func(fun, table),
            Decl::Constant(c) => constant(c, table),
        }
    }
}

fn func(fun: &node::FuncDecl, table: &mut SymbolTable) {
    table.insert(fun.name.clone(), Binding::Function(fun.clone()));
}

fn constant(c: &node::ConstantDecl, table: &mut SymbolTable) {
    table.insert(c.name.clone(), Binding::Constant(c.clone()));
}
