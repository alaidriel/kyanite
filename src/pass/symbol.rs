use std::collections::HashMap;

use crate::ast::{node, File, Node, Type};

#[derive(Debug)]
pub struct Function {
    pub(super) arity: usize,
    pub(super) ty: Type,
    pub(super) params: Vec<(String, Type)>,
}

impl Function {
    pub fn new(arity: usize, ty: Type, params: Vec<(String, Type)>) -> Self {
        Self { arity, ty, params }
    }
}

#[derive(Debug)]
pub enum Symbol {
    Function(Function),
    Variable(Type),
}

impl Symbol {
    pub fn ty(&self) -> Type {
        match self {
            Symbol::Function(f) => f.ty.clone(),
            Symbol::Variable(ty) => ty.clone(),
        }
    }
}

#[derive(Debug)]
pub struct SymbolTable(HashMap<String, Symbol>);

impl SymbolTable {
    pub fn new() -> Self {
        Self(HashMap::new())
    }

    pub fn get(&self, name: &String) -> Option<&Symbol> {
        self.0.get(name)
    }

    pub fn insert(&mut self, name: String, symbol: Symbol) {
        self.0.insert(name, symbol);
    }
}

impl From<Vec<(String, Symbol)>> for SymbolTable {
    fn from(vars: Vec<(String, Symbol)>) -> Self {
        Self(vars.into_iter().collect())
    }
}

impl From<&File> for SymbolTable {
    fn from(program: &File) -> Self {
        let mut table = Self::new();
        for node in &program.nodes {
            self::SymbolTableVisitor::visit(node, &mut table);
        }
        // dbg!(&table);
        table
    }
}

trait SymbolTableVisitor {
    fn visit(&self, table: &mut SymbolTable);
}

impl SymbolTableVisitor for Node {
    fn visit(&self, table: &mut SymbolTable) {
        match self {
            Node::FuncDecl(fun) => func(fun, table),
            Node::ConstantDecl(c) => constant(c, table),
            _ => {}
        }
    }
}

fn func(fun: &node::FuncDecl, table: &mut SymbolTable) {
    let params: Vec<(String, Type)> = fun
        .params
        .iter()
        .map(|param| (String::from(&param.name), Type::from(&param.ty)))
        .collect();
    table.insert(
        String::from(&fun.name),
        Symbol::Function(Function::new(
            fun.params.len(),
            Type::from(fun.ty.as_ref()),
            params,
        )),
    );
}

fn constant(c: &node::ConstantDecl, table: &mut SymbolTable) {
    table.insert(String::from(&c.name), Symbol::Variable(Type::from(&c.ty)));
}
