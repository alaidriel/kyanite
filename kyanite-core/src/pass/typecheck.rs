use std::ops::Deref;

use crate::{
    ast::{node, File, Node, NodeSpan, Type},
    reporting::error::PreciseError,
    token::{Span, Token, TokenKind},
    Source,
};

use super::symbol::SymbolTable;

pub struct TypeCheckPass<'a> {
    source: &'a Source,
    program: &'a File,
    errors: Vec<PreciseError>,
    function: Option<Token>,
    scopes: Vec<SymbolTable>,
}

impl<'a> TypeCheckPass<'a> {
    pub fn new(table: SymbolTable, source: &'a Source, program: &'a File) -> Self {
        Self {
            source,
            program,
            errors: vec![],
            function: None,
            scopes: vec![table],
        }
    }

    pub fn run(&mut self) -> Result<(), usize> {
        for node in &self.program.nodes {
            self.check(node);
        }
        let len = self.errors.len();
        if len > 0 {
            Err(len)
        } else {
            Ok(())
        }
    }

    fn scope_mut(&mut self) -> &mut SymbolTable {
        self.scopes.last_mut().unwrap()
    }

    fn begin_scope(&mut self) {
        self.scopes.push(SymbolTable::new());
    }

    fn end_scope(&mut self) {
        self.scopes.pop();
    }

    fn symbol(&self, name: &Token) -> Option<&Node> {
        for scope in self.scopes.iter().rev() {
            if let Some(definition) = scope.get(name) {
                return Some(definition);
            }
        }
        None
    }

    fn error(&mut self, at: Span, heading: String, text: String) {
        let error = PreciseError::new(self.source, at, heading, text);
        println!("{}", error);
        self.errors.push(error);
    }

    fn check(&mut self, node: &Node) -> Type {
        match node {
            Node::Float(..) => Type::Float,
            Node::Int(..) => Type::Int,
            Node::Bool(..) => Type::Bool,
            Node::Str(..) => Type::Str,
            Node::ConstantDecl(c) => self.constant(c),
            Node::VarDecl(v) => self.var(v),
            Node::FuncDecl(fun) => self.function(fun),
            Node::Return(r) => self.ret(r),
            Node::Binary(b) => self.binary(b),
            Node::Ident(id) => self.ident(id),
            Node::Call(call) => self.call(call),
            e => todo!("{:?}", e),
        }
    }

    fn constant(&mut self, c: &node::ConstantDecl) -> Type {
        let got = self.check(&c.expr);
        let expected = Type::from(&c.ty);
        if got != expected {
            self.error(
                c.expr.span(),
                format!("expected initializer to be of type {}", expected),
                format!("expression of type {}", got),
            );
        }
        expected
    }

    fn var(&mut self, v: &node::VarDecl) -> Type {
        let got = self.check(&v.expr);
        let expected = Type::from(&v.ty);
        if got != expected {
            self.error(
                v.expr.span(),
                format!("expected initializer to be of type {}", expected),
                format!("expression of type {}", got),
            );
        }
        self.scope_mut()
            .insert(v.name.clone(), Node::VarDecl(v.clone()));
        expected
    }

    fn function(&mut self, fun: &node::FuncDecl) -> Type {
        self.begin_scope();
        self.function = Some(fun.name.clone());
        for param in &fun.params {
            self.scope_mut()
                .insert(param.name.clone(), Node::FuncDecl(fun.clone()));
        }
        for node in &fun.body {
            self.check(node);
        }
        self.end_scope();
        self.function = None;
        Type::Void
    }

    fn ret(&mut self, r: &node::Return) -> Type {
        let got = self.check(&r.expr);
        match &self.function {
            Some(function) => {
                let symbol = self.symbol(function).unwrap();
                if got != symbol.ty() {
                    self.error(
                        r.expr.span(),
                        format!("expected return type to be {}", symbol.ty()),
                        format!("expression is of type {}", got),
                    );
                }
            }
            None => unimplemented!("disallowed by parser"),
        }
        got
    }

    fn binary(&mut self, b: &node::Binary) -> Type {
        let lty = self.check(&b.left);
        let rty = self.check(&b.right);
        if lty != rty {
            let heading = match b.op.kind {
                TokenKind::Plus => format!("cannot add {} to {}", lty, rty),
                TokenKind::Minus => format!("cannot subtract {} from {}", rty, lty),
                TokenKind::Star => format!("cannot multiply {} by {}", lty, rty),
                TokenKind::Slash => format!("cannot divide {} by {}", lty, rty),
                _ => format!("cannot compare {} and {}", lty, rty),
            };
            self.error(b.op.span, heading, "".into());
        }
        lty
    }

    fn ident(&mut self, id: &node::Ident) -> Type {
        match self.symbol(&id.name) {
            Some(Node::FuncDecl(f)) => {
                let param = f.params.iter().find(|p| p.name == id.name).unwrap();
                Type::from(&param.ty)
            }
            Some(Node::VarDecl(v)) => Type::from(&v.ty),
            Some(Node::ConstantDecl(c)) => Type::from(&c.ty),
            Some(node) => unimplemented!("{:?} is not implemented as an identifier", node),
            _ => {
                self.error(id.name.span, "is not defined".into(), "".into());
                Type::Void
            }
        }
    }

    fn call(&mut self, call: &node::Call) -> Type {
        let name = match &call.left.deref() {
            Node::Ident(ident) => ident.name.clone(),
            Node::Binary(_) => todo!(),
            _ => unimplemented!(),
        };
        let (arity, params, ty) = match self.symbol(&name) {
            Some(Node::FuncDecl(function)) => (
                function.params.len(),
                function.params.clone(),
                function.ty.clone(),
            ),
            Some(_) => {
                self.error(
                    call.parens.0.span,
                    format!("{} is not a function", name),
                    "".into(),
                );
                return Type::Void;
            }
            _ => {
                let name = String::from(&name);
                // TODO: more robust type checking for builtins
                if name != "println" && name != "max" && name != "min" {
                    self.error(
                        call.parens.0.span,
                        format!("{} is not defined", name),
                        "".into(),
                    );
                }
                return Type::Void;
            }
        };
        if arity != call.args.len() {
            self.error(
                call.left.span(),
                format!(
                    "this function takes {} arguments, but {} were provided",
                    arity,
                    call.args.len()
                ),
                format!("while calling {} here", name),
            );
        }
        for (i, arg) in call.args.iter().enumerate() {
            let got = self.check(arg);
            if i < params.len() {
                let expected = Type::from(&params[i].ty);
                if got != expected {
                    self.error(
                        arg.span(),
                        format!("expected argument of type {}, but found {}", expected, got),
                        format!("expression of type {}", got),
                    );
                }
            }
        }
        if let Some(ty) = ty {
            Type::from(&ty)
        } else {
            Type::Void
        }
    }
}

macro_rules! assert_typecheck {
    ($($path:expr => $name:ident),*) => {
        #[cfg(test)]
        mod tests {
            use crate::{SymbolTable, pass::typecheck::TypeCheckPass};

            $(
                #[test]
                fn $name() -> Result<(), Box<dyn std::error::Error>> {
                    let source = crate::Source::new($path)?;
                    let ast = crate::ast::Ast::from_source(source.clone())?;
                    let symbols = SymbolTable::from(&ast.file);
                    let mut pass = TypeCheckPass::new(symbols, &source, &ast.file);
                    let _ = pass.run();
                    insta::with_settings!({snapshot_path => "../snapshots"}, {
                        insta::assert_yaml_snapshot!(pass.errors);
                    });

                    Ok(())
                }
            )*
        }
    };
}

assert_typecheck! {
    "test-cases/typecheck/varied.kya" => varied
}
