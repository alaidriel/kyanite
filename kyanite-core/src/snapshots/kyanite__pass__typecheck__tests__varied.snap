---
source: kyanite-core/src/pass/typecheck.rs
expression: pass.errors
---
- filename: test-cases/typecheck/varied.kya
  heading: expected initializer to be of type int
  line: "const FOO: int = 5. > 7. * 4.;"
  span:
    line: 1
    column: 18
    length: 12
  text: expression of type float
- filename: test-cases/typecheck/varied.kya
  heading: cannot add float to int
  line: "    return a + b; % mismatched types for binary op '+'"
  span:
    line: 8
    column: 14
    length: 1
  text: ""
- filename: test-cases/typecheck/varied.kya
  heading: cannot subtract float from int
  line: "    return a - b + 7; % mismatched types for binary op '-' and '+'"
  span:
    line: 12
    column: 14
    length: 1
  text: ""
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type int, but found float"
  line: "    return foo(a, b) + bar(c, d); % mismatched types for binary op '+'"
  span:
    line: 20
    column: 31
    length: 1
  text: expression of type float
- filename: test-cases/typecheck/varied.kya
  heading: cannot add int to float
  line: "    return foo(a, b) + bar(c, d); % mismatched types for binary op '+'"
  span:
    line: 20
    column: 22
    length: 1
  text: ""
- filename: test-cases/typecheck/varied.kya
  heading: expected initializer to be of type float
  line: "    let c: float = a * b;"
  span:
    line: 34
    column: 20
    length: 5
  text: expression of type int
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type float, but found int"
  line: "    bar(a, b); % mismatched types 'float' and 'int'"
  span:
    line: 36
    column: 9
    length: 1
  text: expression of type int
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type float, but found int"
  line: "    bar(within(17), 3);"
  span:
    line: 38
    column: 9
    length: 10
  text: expression of type int
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type float, but found int"
  line: "    within(withinFloat(1000));"
  span:
    line: 40
    column: 24
    length: 4
  text: expression of type int
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type int, but found float"
  line: "    within(withinFloat(1000));"
  span:
    line: 40
    column: 12
    length: 17
  text: expression of type float
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type int, but found str"
  line: "    within(withinFloat(within(\"hello\"))); % mismatched types 'float' and 'string'"
  span:
    line: 41
    column: 31
    length: 7
  text: expression of type str
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type float, but found int"
  line: "    within(withinFloat(within(\"hello\"))); % mismatched types 'float' and 'string'"
  span:
    line: 41
    column: 24
    length: 15
  text: expression of type int
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type int, but found float"
  line: "    within(withinFloat(within(\"hello\"))); % mismatched types 'float' and 'string'"
  span:
    line: 41
    column: 12
    length: 28
  text: expression of type float
- filename: test-cases/typecheck/varied.kya
  heading: "this function takes 2 arguments, but 3 were provided"
  line: "    foo(a, b, 10); % mismatched args"
  span:
    line: 43
    column: 5
    length: 3
  text: while calling foo here
- filename: test-cases/typecheck/varied.kya
  heading: "expected argument of type int, but found float"
  line: "    within(baz); % mismatched args"
  span:
    line: 46
    column: 12
    length: 3
  text: expression of type float
- filename: test-cases/typecheck/varied.kya
  heading: cannot add float to int
  line: "    baz + 7;"
  span:
    line: 48
    column: 9
    length: 1
  text: ""
- filename: test-cases/typecheck/varied.kya
  heading: expected return type to be int
  line: "    return 12. * 36.; % mismatched types 'int' and 'float'"
  span:
    line: 50
    column: 12
    length: 9
  text: expression is of type float

